File: impl/ts/tests/operator_cost.test.ts
// Per-operator cost base (functions.md §8). The evaluator charges operatorCost(name) for an
// operator node instead of a flat operatorEval; operatorCost returns the operator's catalog cost if
// authored, else the uniform operatorEval. The conformance corpus CANNOT observe this while every
// built-in uses the uniform default (CLAUDE.md §10), so these per-core tests pin the mechanism.
// Mirrored in Rust (executor.rs registry_tests) and Go (operator_cost_test.go).

import assert from "node:assert/strict";
import { test } from "node:test";
import { COSTS } from "../src/costs.ts";
import { operatorCost } from "../src/executor.ts";
import { OPERATORS } from "../src/operators.ts";

// operatorCost must reflect the generated OPERATORS table for EVERY operator — proving the lookup
// is data-driven, so authoring a cost in catalog.toml is honored with no evaluator change.
test("operatorCost reflects the catalog cost field", () => {
  for (const o of OPERATORS) {
    const want = o.cost === 0 ? COSTS.operatorEval : BigInt(o.cost);
    assert.equal(operatorCost(o.name), want, `operatorCost(${o.name})`);
  }
  // An unknown name falls back to the uniform operatorEval.
  assert.equal(operatorCost("definitely_not_an_operator"), COSTS.operatorEval);
});

// Every operator name the evaluator charges through (arith/comparison ops ARE the catalog names;
// neg/not/and/or are wired literals) must resolve to a real catalog operator.
test("wired operator names exist in the catalog", () => {
  const names = [
    "add",
    "sub",
    "mul",
    "div",
    "mod",
    "eq",
    "ne",
    "lt",
    "gt",
    "le",
    "ge",
    "neg",
    "not",
    "and",
    "or",
  ];
  for (const name of names) {
    assert.ok(
      OPERATORS.some((o) => o.name === name),
      `wired operator name ${name} is not in the catalog`,
    );
  }
});

File: impl/go/operator_cost_test.go
package jed

import "testing"

// Per-operator cost base (functions.md §8). The evaluator charges operatorCost(name) for an
// operator node instead of a flat OperatorEval; operatorCost returns the operator's catalog Cost if
// authored, else the uniform OperatorEval. The conformance corpus CANNOT observe this while every
// built-in uses the uniform default (CLAUDE.md §10), so these per-core tests pin the mechanism.
// Mirrored in Rust (executor.rs registry_tests) and TS (tests/operator_cost.test.ts).

// operatorCost must reflect the generated Operators table for EVERY operator — proving the lookup
// is data-driven, so authoring a Cost in catalog.toml is honored with no evaluator change.
func TestOperatorCostReflectsCatalog(t *testing.T) {
	t.Parallel()
	for _, o := range operators {
		want := costs.OperatorEval
		if o.Cost != 0 {
			want = o.Cost
		}
		if got := operatorCost(o.Name); got != want {
			t.Errorf("operatorCost(%q) = %d, want %d", o.Name, got, want)
		}
	}
	// An unknown name falls back to the uniform OperatorEval.
	if got := operatorCost("definitely_not_an_operator"); got != costs.OperatorEval {
		t.Errorf("unknown operator cost = %d, want %d", got, costs.OperatorEval)
	}
}

// Every operator-enum → catalog-name mapping the evaluator charges through must resolve to a real
// catalog operator, so a typo in catalogName / a wired literal is caught here, not silently masked
// by the uniform-weight fallback.
func TestWiredOperatorNamesExistInCatalog(t *testing.T) {
	t.Parallel()
	var names []string
	for _, op := range []binaryOp{opAdd, opSub, opMul, opDiv, opMod, opEq, opNe, opLt, opGt, opLe, opGe} {
		names = append(names, op.catalogName())
	}
	names = append(names, "neg", "not", "and", "or")
	for _, name := range names {
		found := false
		for _, o := range operators {
			if o.Name == name {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("wired operator name %q is not in the catalog", name)
		}
	}
}
